### DISCUSSION TITLE
Use a library like `typia` for implementing assertions

### DISCUSSION THREAD

**Discussion by brillout** [vikejs/vike#753] [Discussion]
As suggested by @magne4000, we may want to use a library like [typia](https://github.com/samchon/typia) to implement assertions.

Although, so far, I'm happy with the [current implementation](https://github.com/brillout/vite-plugin-ssr/blob/84e7287eedb1c77fdee1d2d0d8e0f9f22d592a84/vite-plugin-ssr/utils/assert.ts) so that's why I'm leaning towards sticking to that implementation.

**Comment by magne4000** [comment]
The cool thing with `typia` is that it inlines generated runtime checks, thus avoiding a lot of potential function calls.
So it can be beneficial on code bases which relies a lot on assert-like function calls.

However, I haven't run any benchmark on any vps part, so it's more like a thing to keep in mind if at some point we see some slowdown/heavy usage on specific parts.

**Reply by brillout** [reply]
I see, makes sense.

**Comment by magne4000** [comment]
I tried `typia` a bit, it's not that great:
- It generates a lot more code than expected
- It generates functions calls
- It doesn't seem tree-shakable
- It doesn't support esbuild/swc/babel, only rollup/vite

**Reply by brillout** [reply]
I wonder how big the performance improvements would be. For frontend frameworks I'm inclined to think that data fetching, and I/O in general, is that much more of a bottleneck to the point that we can neglect optimizing the JavaScript execution itself. Although I can see that in some parts of the code, such as stream handling, it could very much have a non-negligible impact.

One thing to note: having the `assert()` calls in production is very convenient for fixing bugs. I'd even say that catching bugs early is a reason why VPS is bug free. There were many situations in the past where, thanks to the early assertion failure, the fix was straigthforward whereas letting the app crash would have been a *lot* more time consuming to fix.
